### PR TITLE
Add MLPerf RCP logging for post training in DeepSWE

### PR DESCRIPTION
**Pull request by susanbao**
Resolves #

### Description
This PR adds MLPerf RCP logging utilities and integration for post-training (GRPO) in DeepSWE:

- **MLPerf Logging Utilities** (`examples/deepswe/mllog_utils.py`): Implements MLPerf logging lifecycle hooks (`init_start`, `init_stop`, `run_start`, `block_start`, `block_stop`, `init_print`, `start_eval`, `end_eval`, `check_eval`, `log_tracked_stats`, `run_stop`) with multi-host coordination, file output configuration, and fallback handling when `mlperf_logging` is unavailable.
- **DeepSWE Integration** (`examples/deepswe/train_deepswe_nb.py`): Adds `--rcp_logging` flag to enable MLPerf initialization, submission metadata logging, and run lifecycle events.
- **Agentic RL Learner Integration** (`tunix/rl/agentic/agentic_rl_learner.py`): Logs per-step training metrics (losses, rewards, step times, advantages, completion lengths, and actor metrics) under `rcp_logging` with safe optional imports and proper step indexing.
- **Unit & Compliance Tests** (`tests/examples/deepswe_mllog_utils_test.py`): Adds tests covering file logging, eval lifecycle, early stopping, and end-to-end submission config logging.

> **Note**: The eval logging is not added to `agentic_rl_learner.py` as the eval pipeline has not validated with DeepSWE now.

**Reference**
Tested RCP logging with Qwen3.5 35B. After mocking eval and part of parameters (e.g. submission_benchmark), the logging passed the MLPerf ruling check:
```
INFO - Running compliance on file: corrected_sb_35b_20_mlperf_compliance.log
INFO - Compliance checks: training_6.1.0/closed_qwen35_397b_grpo.yaml
** Logging output also at compliance_checker.log
INFO - SUCCESS
```

**Colab Notebook**
N/A

**Checklist**

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and all unit tests pass.
- [x] I have added all appropriate doc-strings/documentation.
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).
- [x] I have followed [Contribution Guidelines](https://github.com/google/tunix/blob/main/docs/contributing.md).

### CODE REVIEW

File: tunix/rl/agentic/agentic_rl_learner.py [hunk 1]
```diff
@@ -53,6 +53,11 @@
 from tunix.rl.queue import data_queue as queue_lib
 from tunix.sft import utils as sft_utils
 
+try:
+  from examples.deepswe import mllog_utils
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_rl_learner.py]: hmmm...this looks pretty anti-pattern...can we not import from examples in the core training component?

**Review by tianshub — CHANGES_REQUESTED** [review]

File: tunix/utils/mllog_utils.py [hunk 2]
```diff
@@ -0,0 +1,483 @@
+# Copyright 2026 Google LLC
```
Comment (tianshub) [inline tunix/utils/mllog_utils.py:1]: I'm missing some context what's the point of the RCP logging, are you trying to add a custom metric logger? if so we have a public API here that you can wire in much easily: https://github.com/google/tunix/blob/cad21c9d21f981fde8b023ae927dcfd2c62172c0/tunix/rl/rl_cluster.py#L642

File: examples/deepswe/train_deepswe_nb.py [hunk 3]
```diff
@@ -1060,6 +1068,7 @@ def filter_logical_rules(rules, mesh):
     "loss_agg_mode": LOSS_AGG_MODE,
     "advantage_estimator": ADVANTAGE_ESTIMATOR,
     "use_rollout_logps": USE_ROLLOUT_LOGPS,
+    "rcp_logging": args.rcp_logging,
```
Comment (RissyRan) [inline examples/deepswe/train_deepswe_nb.py]: Nit, reuse `RCP_LOGGING`

File: examples/deepswe/train_deepswe_nb.py [hunk 4]
```diff
@@ -1116,8 +1125,26 @@ def filter_logical_rules(rules, mesh):
   print(f"W&B initialization failed with error: {e}")
 
 
+if RCP_LOGGING:
```
Comment (RissyRan) [inline examples/deepswe/train_deepswe_nb.py:1138]: Could you create a bug to track the RCP logging work for openhands once we have it?

File: tunix/rl/agentic/agentic_rl_learner.py (hunk 1 above)
Comment (RissyRan) [reply tunix/rl/agentic/agentic_rl_learner.py]: Let's move the mllog_utils file under `tunix/utils`?

https://github.com/google/tunix/tree/7a657b25f9d07c49946c1c57f0ecf00d84208f45/tunix/utils

**Review by RissyRan — COMMENTED** [review]
Thanks Sanbao!

I didn't see detailed metrics are reporting, like exposed_generation_time, train_step_time, policy_training_time, etc (only in test). Will you integrate them as next step?

File: tunix/rl/agentic/agentic_rl_learner.py (hunk 1 above)
Comment (susanbao) [reply tunix/rl/agentic/agentic_rl_learner.py]: I have used with_external_metrics_logger to avoid import the mllog.

File: examples/deepswe/train_deepswe_nb.py (hunk 4 above)
Comment (susanbao) [reply examples/deepswe/train_deepswe_nb.py:1138]: Have created b/553025035.

File: tunix/utils/mllog_utils.py (hunk 2 above)
Comment (susanbao) [reply tunix/utils/mllog_utils.py:1]: Have used this public API.

File: examples/deepswe/train_deepswe_nb.py (hunk 3 above)
Comment (susanbao) [reply examples/deepswe/train_deepswe_nb.py]: Have fixed. These metrics are reported. Only eval part is not as we have not validated the eval pipeline now.

File: tunix/rl/agentic/agentic_rl_learner.py [hunk 5]
```diff
@@ -1124,8 +1128,22 @@ async def _eval_runner_async(current_eval_orchestrator):
             eval_str,
         )
         did_eval_this_global_step = False
+        step_metrics = {
+            "perf/global_step_time": (global_step_time, np.mean),
+            "step_time": (global_step_time, np.mean),
+        }
+        if train_rewards.size:
```
Comment (tianshub) [inline tunix/rl/agentic/agentic_rl_learner.py]: @susanbao I believe we are already logging the global_step_time, reward and generation length metrics, is there any missing metrics you see?

File: examples/deepswe/train_deepswe_nb.py [hunk 6]
```diff
@@ -1039,6 +1047,11 @@ def filter_logical_rules(rules, mesh):
     cluster_config=cluster_config,
 )
 
+if RCP_LOGGING:
```
Comment (tianshub) [inline examples/deepswe/train_deepswe_nb.py:1049]: thanks for taking the suggestion, it looks much cleaner now.

File: examples/deepswe/train_deepswe_nb.py [hunk 7]
```diff
@@ -1116,8 +1129,30 @@ def filter_logical_rules(rules, mesh):
   print(f"W&B initialization failed with error: {e}")
 
 
+if RCP_LOGGING:
+  mllog_utils.init_print(
```
Comment (tianshub) [inline examples/deepswe/train_deepswe_nb.py]: this can be moved to L1053?

File: examples/deepswe/train_deepswe_nb.py [hunk 8]
```diff
@@ -1116,8 +1129,30 @@ def filter_logical_rules(rules, mesh):
...
+      args,
+      train_dataset=dataset,
+      rollout_mesh=rollout_mesh,
+      train_mesh=train_mesh,
+      total_devices=total_devices,
+  )
+  mllog_utils.init_stop()
```
Comment (tianshub) [inline examples/deepswe/train_deepswe_nb.py]: can these be merged into a single API?

File: examples/deepswe/train_deepswe_nb.py [hunk 9]
```diff
@@ -1116,8 +1135,21 @@ def filter_logical_rules(rules, mesh):
...
 
 
+if RCP_LOGGING:
+  mllog_utils.train_start(args)
+
 print("Starting training...", flush=True)
 agentic_grpo_learner.train(train_dataset=train_dataset)
 
+if RCP_LOGGING:
+  mllog_utils.block_stop(
```
Comment (tianshub) [inline examples/deepswe/train_deepswe_nb.py]: these two can be merged into one API as well?

**Review by tianshub — APPROVED** [review]
